src/core/qgstextrenderer.cpp

```diff
@@ -2559,7 +2579,24 @@ void QgsTextRenderer::drawBuffer( QgsRenderContext &context, const QgsTextRender
 
   QPainterPath path;
   path.setFillRule( Qt::WindingFill );
-  path.addText( 0, 0, format.scaledFont( context ), component.text );
+  switch ( format.orientation() )
+  {
+    case QgsTextFormat::HorizontalOrientation:
+      path.addText( 0, 0, format.scaledFont( context ), component.text );
+      break;
+    case QgsTextFormat::VerticalOrientation:
+      double letterSpacing = format.scaledFont( context ).letterSpacing();
+      double labelWidth = fontMetrics->maxWidth();
+      const QStringList parts = QgsPalLabeling::splitToGraphemes( component.text );
+      double partYOffset = 0.0;
+      for ( const auto &part : parts )
+      {
+        double partXOffset = ( labelWidth - ( fontMetrics->width( part ) - letterSpacing ) ) / 2;
+        path.addText( partXOffset, partYOffset, format.scaledFont( context ), part );
+        partYOffset += fontMetrics->descent() + fontMetrics->ascent() + letterSpacing;
```

Comment:
Just wondering about this. I realise it's correct logic for scripts with equal character heights, but for others (eg Latin) shouldn't we consider the actual character heights?

Reply:
I'd rather follow the decisions by much wiser people who codified the behavior in firefox, chrome, et cie and stick to fixed height. 

For one, we need a perfect alignment of characters across each lines all eastern scripts. But its also the case that latin characters would be far less readable with variable heights. Dots would become a tiny... dot and would merge with other characters etc.